### PR TITLE
Update spec for test changes, add "bench" subcommand

### PR DESCRIPTION
**Pull request by paf31**
Updates ndc-spec and adds a bench subcommand for simple benchmarks of connectors built with the SDK:

```
Usage: ndc_hub_example bench [OPTIONS] --configuration <DIRECTORY> --snapshots-dir <DIRECTORY>

Options:
      --configuration <DIRECTORY>  [env: HASURA_CONFIGURATION_DIRECTORY=]
      --samples <COUNT>            the number of samples to collect per test [default: 100]
      --tolerance <TOLERANCE>      tolerable deviation from previous report, in standard deviations from the mean
      --snapshots-dir <DIRECTORY>  [env: HASURA_SNAPSHOTS_DIR=]
  -h, --help                       Print help
```

### CODE REVIEW

File: rust-connector-sdk/src/default_main.rs [hunk 1]
```diff
@@ -531,17 +531,20 @@ async fn test<Setup: ConnectorSetup>(
...
+    let mut reporter = (ConsoleReporter::new(), TestResults::default());
 
-    if !results.failures.is_empty() {
-        println!();
-        println!("{}", report(results));
+    ndc_test::test_connector(&test_configuration, &connector, &mut reporter).await;
+
+    if !reporter.1.failures.is_empty() {
+        // println!();
+        // println!("{}", report(reporter.1));
```
[the remark this answers is not in the document]
Comment (paf31) [reply rust-connector-sdk/src/default_main.rs:549]: https://github.com/hasura/ndc-spec/pull/106

File: rust-connector-sdk/Cargo.toml [hunk 2]
```diff
@@ -13,8 +13,8 @@ path = "bin/main.rs"
 
 [dependencies]
 gdc_rust_types = { git = "https://github.com/hasura/gdc_rust_types.git", rev = "3273434" }
-ndc-client = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.18" }
-ndc-test = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.18" }
+ndc-client = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.19" }
+ndc-test = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.19" }
```
Comment (SamirTalwar) [inline rust-connector-sdk/Cargo.toml:17]: I think that in the future, we should consider moving the `ndc-test`-related stuff to a separate, optional crate.

I want this testing and benchmarking but I don't want to ship it to prod.

FWIW we don't use the `test` command in ndc-postgres; we wrap the test runner in an integration test.

File: Cargo.lock
Comment (SamirTalwar) [inline Cargo.lock:1]: Love to see it.

File: rust-connector-sdk/src/default_main.rs [hunk 3]
```diff
@@ -531,17 +557,20 @@ async fn test<Setup: ConnectorSetup>(
...
-    let results = ndc_test::test_connector(&test_configuration, &connector).await;
+    let mut reporter = (ConsoleReporter::new(), TestResults::default());
 
-    if !results.failures.is_empty() {
+    ndc_test::test_connector(&test_configuration, &connector, &mut reporter).await;
+
+    if !reporter.1.failures.is_empty() {
         println!();
-        println!("{}", report(results));
+        println!("{}", reporter.1.report());
```
Comment (SamirTalwar) [inline rust-connector-sdk/src/default_main.rs:573]: Why is this necessary? Doesn't the `ConsoleReporter` handle this now?

(Same applies to the `replay` function.)

**Review by SamirTalwar — APPROVED** [review]

File: rust-connector-sdk/src/default_main.rs (hunk 3 above)
Comment (paf31) [reply rust-connector-sdk/src/default_main.rs:573]: The reporter only responds to the traversal of the nodes of the test tree structure, it doesn't have a hook for when the tests are fully complete. `ConsoleReporter` writes the tree out to stdout without allocating, but `TestResults` collects up the failures. Another approach would be to add a `finish` method to the trait, but as it is, `TestResults` is useful for cases where you don't want to write to stdout at all, or immediately.

File: rust-connector-sdk/Cargo.toml (hunk 2 above)
Comment (paf31) [reply rust-connector-sdk/Cargo.toml:17]: We could put it behind a feature flag.

Comment (paf31) [reply rust-connector-sdk/Cargo.toml:17]:
```
% cargo run
Usage: ndc_hub_example <COMMAND>

Commands:
  serve         
  check-health  
  help          Print this message or the help of the given subcommand(s)

% cargo run --features ndc-test

Usage: ndc_hub_example <COMMAND>

Commands:
  serve         
  test          
  replay        
  bench         
  check-health  
  help          Print this message or the help of the given subcommand(s)
```

Comment (SamirTalwar) [reply rust-connector-sdk/Cargo.toml]: Yeah, that's what I was thinking, but I'm fine with doing it later.

File: rust-connector-sdk/Cargo.toml [hunk 4]
```diff
@@ -11,10 +11,14 @@ path = "src/lib.rs"
...
+[features]
+ndc-test = ["dep:ndc-test"]
+default = ["ndc-test"]
+
 [dependencies]
 gdc_rust_types = { git = "https://github.com/hasura/gdc_rust_types.git", rev = "3273434" }
-ndc-client = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.18" }
-ndc-test = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.18" }
+ndc-client = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.19" }
+ndc-test = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0-rc.19", optional = true }
```
Comment (SamirTalwar) [inline rust-connector-sdk/Cargo.toml:21]: I think we can use v0.1.0 now!

```suggestion
ndc-client = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0" }
ndc-test = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.0", optional = true }
```

**Review by SamirTalwar — APPROVED** [review]
Let's ship this!
